### PR TITLE
[improvement](pipeline) task group scan entity

### PR DESCRIPTION
**Pull request by liutang123**
scan task group queue

# Proposed changes

Issue Number: close #xxx

## Problem summary

1. Change task group entity by template.
2. add scan task queue and support task group
3. Schedule scan task in `scan task scheduler` if workload group is set.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by liutang123** [comment]
run buildall

**Comment by liutang123** [comment]
run buildall

**Comment by liutang123** [comment]
run p0

File: be/src/pipeline/task_scheduler.h [hunk 1]
```diff
@@ -91,8 +91,7 @@ class TaskScheduler {
 
     void shutdown();
 
-    void update_tg_cpu_share(const taskgroup::TaskGroupInfo& task_group_info,
-                             taskgroup::TaskGroupPtr task_group);
+    TaskQueue* task_queue() const { return _task_queue.get(); }
```
Comment (wangbo) [inline be/src/pipeline/task_scheduler.h:94]: _task_queue is shared by task_scheduler and TaskGroup, so it' better use shared_ptr here.

File: be/src/vec/exec/scan/scan_task_queue.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+public:
+    explicit ScanTaskTaskGroupQueue(size_t core_size);
+    ~ScanTaskTaskGroupQueue();
+
+    void close();
+    bool take(ScanTask* scan_task);
+    template <TabletStorageType storage_type>
+    bool push_back(ScanTask);
+
+    // ScanTask可以用ScannerContext代替
```
Comment (wangbo) [inline be/src/vec/exec/scan/scan_task_queue.h:71]: In english

File: be/src/vec/exec/scan/scan_task_queue.cpp [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    }
+}
+
+template TGSTEntityPtr ScanTaskTaskGroupQueue::_task_entity<TabletStorageType::STORAGE_TYPE_LOCAL>(
+        ScanTask& scan_task);
+template TGSTEntityPtr ScanTaskTaskGroupQueue::_task_entity<TabletStorageType::STORAGE_TYPE_REMOTE>(
+        ScanTask& scan_task);
+template <TabletStorageType storage_type>
+TGSTEntityPtr ScanTaskTaskGroupQueue::_task_entity(ScanTask& scan_task) {
+    auto* group = scan_task.scanner_context->get_task_group();
```
Comment (wangbo) [inline be/src/vec/exec/scan/scan_task_queue.cpp:123]: Maybe we can move this branch(local or remote) to scan_task

File: be/src/runtime/task_group/task_group.cpp [hunk 4]
```diff
@@ -40,50 +42,89 @@ const static std::string CPU_SHARE = "cpu_share";
...
+template <typename QueueType, typename ParentQueueType>
+size_t TaskGroupEntity<QueueType, ParentQueueType>::task_size() const {
+    return _task_queue->size();
 }
 
-uint64_t TaskGroupEntity::cpu_share() const {
-    return _tg->cpu_share();
+template <typename QueueType, typename ParentQueueType>
+uint64_t TaskGroupEntity<QueueType, ParentQueueType>::cpu_share() const {
+    return _cpu_share;
```
Comment (wangbo) [inline be/src/runtime/task_group/task_group.cpp:84]: Not need a lock here?

File: be/src/runtime/task_group/task_group.h [hunk 5]
```diff
@@ -31,46 +30,75 @@
...
-    // TODO pipeline use MLFQ
-    std::queue<pipeline::PipelineTask*> _queue;
-    taskgroup::TaskGroup* _tg;
+    QueueType* _task_queue;
+
     uint64_t _vruntime_ns = 0;
+    taskgroup::TaskGroup* _tg;
+
+    int64_t _version;
+    uint64_t _cpu_share;
```
Comment (wangbo) [inline be/src/runtime/task_group/task_group.h:91]: Add a comment to explain why TaskGroupEntity need store a _cpu_share when group has kept cpu_share

**Comment by liutang123** [comment]
run buildall

**Review by wangbo — DISMISSED** [review]
LGTM

**Comment by liutang123** [comment]
run buildall

**Comment by liutang123** [comment]
run buildall

**Comment by liutang123** [comment]
run buildall

**Comment by liutang123** [comment]
run buildall

**Comment by liutang123** [comment]
@wangbo do you have time to merge this pr？

**Review by wangbo — DISMISSED** [review]
LGTM

**Comment by liutang123** [comment]
run buildall

**Review by wangbo — APPROVED** [review]
LGTM

**Review by luozenglin — APPROVED** [review]
LGTM

**Pull request merged**